src/test/java/com/karankumar/bookproject/ui/book/BookFormTest.java

```diff
@@ -353,6 +355,73 @@ void correctFormFieldsShowForDidNotFinishShelf() {
         Assertions.assertTrue(bookForm.pagesReadFormItem.isVisible());
     }
 
+    @Test
+    void fieldsToResetIsCorrectlyPopulatedForToReadShelf() {
+        // given
+        HasValue[] fieldsThatShouldBeReset = new HasValue[]{bookForm.dateStartedReading,
+                bookForm.dateFinishedReading, bookForm.pagesRead, bookForm.rating, bookForm.bookReview};
+        populateBookForm(READ, false);
+        bookForm.pagesRead.setValue(pagesRead);
+
+        // when
+        bookForm.predefinedShelfField.setValue(TO_READ);
+
+        // then
+        Assertions.assertEquals(fieldsThatShouldBeReset.length, bookForm.fieldsToReset.length);
+        Assertions.assertTrue(List.of(bookForm.fieldsToReset)
+                              .containsAll(List.of(fieldsThatShouldBeReset)));
+    }
+
+    @Test
+    void fieldsToResetIsCorrectlyPopulatedForReadingShelf() {
+        // given
+        HasValue[] fieldsThatShouldBeReset = new HasValue[]{bookForm.dateFinishedReading,
+                bookForm.pagesRead, bookForm.rating, bookForm.bookReview};
+        populateBookForm(READ, false);
+        bookForm.pagesRead.setValue(pagesRead);
+
+        // when
+        bookForm.predefinedShelfField.setValue(READING);
+
+        // then
+        Assertions.assertEquals(fieldsThatShouldBeReset.length, bookForm.fieldsToReset.length);
+        Assertions.assertTrue(List.of(bookForm.fieldsToReset)
+                              .containsAll(List.of(fieldsThatShouldBeReset)));
+    }
+
+    @Test
+    void fieldsToResetIsCorrectlyPopulatedForReadShelf() {
+        // given
+        HasValue[] fieldsThatShouldBeReset = new HasValue[]{bookForm.pagesRead};
+        populateBookForm(READ, false);
+        bookForm.pagesRead.setValue(pagesRead);
+
+        // when
+        bookForm.predefinedShelfField.setValue(READ);
+
+        // then
+        Assertions.assertEquals(fieldsThatShouldBeReset.length, bookForm.fieldsToReset.length);
+        Assertions.assertTrue(List.of(bookForm.fieldsToReset)
+                              .containsAll(List.of(fieldsThatShouldBeReset)));
+    }
+
+    @Test
+    void fieldsToResetIsCorrectlyPopulatedForDidNotFinishShelf() {
+        // given
+        HasValue[] fieldsThatShouldBeReset = new HasValue[]{bookForm.dateFinishedReading,
+                bookForm.rating, bookForm.bookReview};
+        populateBookForm(READ, false);
+        bookForm.pagesRead.setValue(pagesRead);
+
+        // when
+        bookForm.predefinedShelfField.setValue(DID_NOT_FINISH);
+
+        // then
+        Assertions.assertEquals(fieldsThatShouldBeReset.length, bookForm.fieldsToReset.length);
+        Assertions.assertTrue(List.of(bookForm.fieldsToReset)
+                              .containsAll(List.of(fieldsThatShouldBeReset)));
+    }
```

Comment:
There's a lot of duplication here. It seems the only things that change are the fieldsThatShouldBeReset and the when statement. Instead, could you change this to be a parameterized test?

Reply:
Also, what do you think about adding four separate reset array fields in `BookForm` for each predefined shelf (e.g. `toReadFieldsToReset`, `readingFieldsToReset`, etc.). Then, you can set `fieldsToReset` to equal the corresponding reset array?

The point of this is so that we can have these four new reset arrays as global variables that are visible for testing. If, at a later date, we want to change one of the fields to reset for a specific predefined shelf, we only need to make that change in one place rather than two.

Reply:
You are right, there really is a lot of duplicated code, I will rewrite the test.

The approach with global arrays also makes sense to me.

Reply:
I had to do a bit of a workaround because I could not use the arrays in the method source, but the test is still a lot cleaner, I think.

Reply:
Thanks